README.md
{\rtf1\ansi\ansicpg1252\cocoartf1404\cocoasubrtf470
{\fonttbl}
{\colortbl;\red255\green255\blue255;}
\paperw11900\paperh16840\margl1440\margr1440\vieww10800\viewh8400\viewkind0
}

COMP47360: RESEARCH PRACTICUM
TEAM 4: OCCUPANTS
BARRY KIDNEY, JOAN MCCARTHY, DAVID O'KEEFFE & ELLEN RUSHE

SERVER: http://csi420-01-vm4.ucd.ie/
LOGIN CREDENTIALS: 
    USERNAME: occupants
    PASSWORD: Research2016!

# REPOSITORY CONTENTS
    DATABASE: All python files to construct the database
    Modelling Notebooks: All iPython notebooks detainling all stages of the modelling process
    Project File: Web Application - This is the final product that has been deployed on the server
    Signal Strength_RaspberryPi: iPython notebooks detailing additional work undertaken in researching the relation between RSSI and the number of people in a room
    Survey: iPython notebooks detailing analysis undertaken on the repsonses received from the survey that was conducted
    Working_Prototype_Presentation2: Early application prototype presented at Presentation 2 'Current Working Prototype'


# USER INSTRUCTIONS
When the user accesses the homepage summary graphs of space utilisation, frequency and occupancy rates are presented to them. Users are also presented with a map of the location of the Computer Science building within UCD. Users also have the option to participate in our survey to help develop the system.  

Users can access the results tab to view historical occupancy results estimated by three Logistic Regression models. Users can select the required room and date to view these results. A graph provides a visualisation of the results and panels alongside the graph provide more detailed results from the models. 

Lecturers and administrative staff have the option to sign up on the navigation bar. Once a new user has received a confirmation email approving their registration the user will then be able to login and to view current data within the database and to upload new data - ground truth, wifi logs, new rooms, timetables and modules. 

# DEVELOPER INSTRUCTIONS
The  application backend is written in [Python 3.5](https://docs.python.org/3.4/). [MySQL](https://www.mysql.com/) is used to provide storage of data. The [Django](https://www.djangoproject.com/) framework for web application development. [WSGI](https://wsgi.readthedocs.org/en/latest/) for providing the communications interface between the web app and Python backend. [Apache2](https://httpd.apache.org/) to host the webpages and web app.

The web application frontend is written in [HTML5](https://www.w3.org/TR/html5/) and [Javascript](http://www.ecma-international.org/publications/standards/Ecma-262.htm). [ChartJS](http://www.chartjs.org/) is used to provide visualisations of results.

conda-meta/history
==> 2016-06-30 14:41:13 <==
# cmd: /home/ellen/anaconda3/bin/conda create --name research django numpy matplotlib scipy pandas scikit-learn
cycler-0.10.0-py35_0
django-1.9.5-py35_0
fontconfig-2.11.1-6
freetype-2.5.5-1
libgfortran-3.0.0-1
libpng-1.6.22-0
libxml2-2.9.2-0
matplotlib-1.5.1-np111py35_0
mkl-11.3.3-0
numpy-1.11.1-py35_0
openssl-1.0.2h-1
pandas-0.18.1-np111py35_0
pip-8.1.2-py35_0
pyparsing-2.1.4-py35_0
pyqt-4.11.4-py35_3
python-3.5.1-5
python-dateutil-2.5.3-py35_0
pytz-2016.4-py35_0
qt-4.8.7-3
readline-6.2-2
scikit-learn-0.17.1-np111py35_2
scipy-0.17.1-np111py35_1
setuptools-23.0.0-py35_0
sip-4.16.9-py35_0
six-1.10.0-py35_0
sqlite-3.13.0-0
tk-8.5.18-0
wheel-0.29.0-py35_0
xz-5.2.2-0
zlib-1.2.8-3
# create specs: ['django', 'numpy', 'matplotlib', 'scipy', 'pandas', 'scikit-learn']
==> 2016-06-30 14:45:30 <==
# cmd: /home/ellen/anaconda3/envs/research/bin/conda install nltk
+nltk-3.2.1-py35_0
# install specs: ['nltk', 'python 3.5* (target=python-3.5.1-5.tar.bz2)']
==> 2016-06-30 14:50:39 <==
# cmd: /home/ellen/anaconda3/envs/research/bin/conda remove --name research sqlite
-cycler-0.10.0-py35_0
-django-1.9.5-py35_0
-matplotlib-1.5.1-np111py35_0
-nltk-3.2.1-py35_0
-numpy-1.11.1-py35_0
-pandas-0.18.1-np111py35_0
-pip-8.1.2-py35_0
-pyparsing-2.1.4-py35_0
-pyqt-4.11.4-py35_3
-python-3.5.1-5
-python-dateutil-2.5.3-py35_0
-pytz-2016.4-py35_0
-scikit-learn-0.17.1-np111py35_2
-scipy-0.17.1-np111py35_1
-setuptools-23.0.0-py35_0
-sip-4.16.9-py35_0
-six-1.10.0-py35_0
-sqlite-3.13.0-0
-wheel-0.29.0-py35_0
# remove specs: ['sqlite']
==> 2016-07-01 10:07:02 <==
# cmd: /home/ellen/anaconda3/envs/research/bin/conda install django
+django-1.9.5-py35_0
+pip-8.1.2-py35_0
+python-3.5.1-5
+setuptools-23.0.0-py35_0
+sqlite-3.13.0-0
+wheel-0.29.0-py35_0
# install specs: ['django']
==> 2016-07-01 10:07:55 <==
# cmd: /home/ellen/anaconda3/envs/research/bin/conda install numpy matplotlib pandas scikit-learn
+cycler-0.10.0-py35_0
+matplotlib-1.5.1-np111py35_0
+numpy-1.11.1-py35_0
+pandas-0.18.1-np111py35_0
+pyparsing-2.1.4-py35_0
+pyqt-4.11.4-py35_3
+python-dateutil-2.5.3-py35_0
+pytz-2016.4-py35_0
+scikit-learn-0.17.1-np111py35_2
+scipy-0.17.1-np111py35_1
+sip-4.16.9-py35_0
+six-1.10.0-py35_0
# install specs: ['numpy', 'matplotlib', 'pandas', 'scikit-learn', 'python 3.5* (target=python-3.5.1-5.tar.bz2)']


index.php
<!DOCTYPE html>
    <!--most code is original but some has been sourced and adapted from www.w3schools.com-->

    <?php
        require_once 'config.php';?><!--import config file, connects to database-->

<html>
    <head> 
        <meta charset="utf-8">
<!--        <link rel="stylesheet" type="text/css" href="css/stylesheet.css">-->
        <link rel="icon" href="/t4cms/ucdmaincore_favicon.ico">
        <link href="https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700,800|Roboto+Slab:100,200,300,400" rel="stylesheet" type="text/css">
        <title>University College Dublin</title>
    </head>
    
    <header><!--import header file, standard header on each page-->
        <?php 
            include 'header.php';?>
    </header>
    
    <body>
        <div class="centered" style="display: block; margin:0px auto 15px; padding: 5px 0px 5px 0px; height:100px;">  
            <form action="<?php echo htmlspecialchars($_SERVER["PHP_SELF"]);?>" method="post">

                <select name="example">
                    <option selected="selected" value="B-002">B-002</option>
                    <option value="B-003">B-003</option>
                    <option value="B-004">B-004</option>
                </select>
                
                <br>
                
                <select name="example3">
                    <option selected="selected" value="2015-11-03">2015-11-03</option>
                    <option value="2015-11-04">2015-11-04</option>
                    <option value="2015-11-05">2015-11-05</option>
                    <option value="2015-11-06">2015-11-06</option>
                    <option value="2015-11-07">2015-11-07</option>
                    <option value="2015-11-08">2015-11-08</option>
                    <option value="2015-11-09">2015-11-09</option>
                    <option value="2015-11-10">2015-11-10</option>
                    <option value="2015-11-11">2015-11-11</option>
                    <option value="2015-11-12">2015-11-12</option>
                    <option value="2015-11-13">2015-11-13</option>
                    <option value="2015-11-14">2015-11-14</option>
                    <option value="2015-11-15">2015-11-15</option>
                    <option value="2015-11-16">2015-11-16</option>
                    <option value="2015-11-17">2015-11-17</option>
                </select>
                
                <input name="submit" type="submit" />
            </form>   
        </div>
            
                <!--setting background image and overlay div-->
        <div style='background-image: url(images/UCDbackground.jpg); background-repeat: no-repeat; background-size: 100%'>
            <div class='centered' style='background-color: #f9f9f5; padding: 20px; opacity: 0.9; filter: alpha(opacity=90); max-width: 1000px; min-height: 550px; margin:0px auto 0px'>
        
        
            <?php
                
                function AvgDevices($arr, $time, $date) {
                    $avgDevs = 0;
                    $counter = 0;

                    foreach($arr as $selected) {
                        if ($selected['Date'] == $date and strtotime($selected['Time']) >= $time+(60*15) and strtotime($selected['Time']) <= $time+(60*45)) {
                            $avgDevs = $avgDevs + $selected['Associated'];
                            $counter = $counter + 1; 
                        }
                    } if ($counter > 0) {
                        return $avgDevs/$counter;
                    } else {
                        return 0;
                    }
                }
                
                
                
                if (isset($_POST['submit']) && !empty($_POST['submit'])) {
                    
                    $example = $_POST['example'];
                    $requestedDate = $_POST['example3'];
                    $requestedCampus = 'Belfield';
                    $requestedBuilding = 'Computer Science';
                    
                    echo $example . "<br>";
                    echo $requestedDate . "<br>";
                    echo $requestedCampus . "<br>";
                    echo $requestedBuilding . "<br>";


                    

                    try { 
                        // construct query
                        $reply = $conn->prepare("SELECT Associated, Date, Time FROM wifilogdata WHERE Room = '{$example}' AND Time >= '09:00:00' AND Time <= '18:00:00'");                        
                        
                        // query the database
                        $reply->execute();
                
                        //error case
                        if(!$reply) {
                            die("Execute query error");
//                          die("Execute query error, because: ". $conn->errorInfo());
                        
                        } else {
                            $array = $reply->fetchAll(PDO::FETCH_ASSOC);

                            // beginning of table 
                            echo "<table style='opacity: 1.0; filter: alpha(opacity=100);'><div><th width='25%'>Time</th><th width='25%'>Devices</th><th width='25%'>Estimated&nbsp;Occupants</th>";
                            
                            // populate the table
                            for($z=strtotime('09:00:00'); $z<strtotime('09:00:00')+(60*60*9); $z+=(60*60)) {
                                $devices = AvgDevices($array, $z, $requestedDate);                
                                $estimate = round($devices/1.27);
                                echo "<tr><td width='25%'>".date("H:i:s", $z)."</td><td width='25%'>".$devices."</td><td width='25%'>".$estimate."</td></tr>";
                            }
                        }
                    } catch(PDOException $e) {
                        echo "Error".$e->getMessage();
                    }
                
                $conn = null;
                // close table
                echo"</table>";
                
                } else {
                    echo "submit not registered.";
                }?>
            </div>
        </div>
    </body>
    
    
    
    <!--import footer file, standard footer on each page-->
    <footer>
        <?php 
            include 'footer.php';
        ?>
    </footer>  
    
</html>